test/WebJobs.Script.Tests/WebScriptHostManagerTests.cs

```diff
@@ -124,6 +125,48 @@ public void SecretFilesArePurgedOnStartup()
         }
 
         [Fact]
+        public async Task MultipleHostRestarts()
+        {
+            string functionTestDir = Path.Combine(_fixture.TestFunctionRoot, Guid.NewGuid().ToString());
+            Directory.CreateDirectory(functionTestDir);
+            string logDir = Path.Combine(_fixture.TestLogsRoot, Guid.NewGuid().ToString());
+            string secretsDir = Path.Combine(_fixture.TestSecretsRoot, Guid.NewGuid().ToString());
+
+            ScriptHostConfiguration config = new ScriptHostConfiguration
+            {
+                RootLogPath = logDir,
+                RootScriptPath = functionTestDir,
+                FileLoggingMode = FileLoggingMode.Always,
+                RestartInterval = TimeSpan.FromMilliseconds(500)
+            };
+            SecretManager secretManager = new SecretManager(secretsDir);
+            WebHostSettings webHostSettings = new WebHostSettings();
+            var factoryMock = new Mock<IScriptHostFactory>();
+            int count = 0;
+            factoryMock.Setup(p => p.Create(config)).Callback(() =>
+            {
+                count++;
+            }).Throws(new Exception("Kaboom!"));
+
+            ScriptHostManager hostManager = new WebScriptHostManager(config, secretManager, webHostSettings, factoryMock.Object);
+
+            Task runTask = Task.Run(() => hostManager.RunAndBlock());
+
+            await TestHelpers.Await(() =>
+            {
+                return count > 3;
+            });
+
+            hostManager.Stop();
+            Assert.False(hostManager.IsRunning);
+
+            // regression test: previously on multiple restarts we were recomposing
+            // the writer on each restart, resulting in a nested chain of writers
+            // increasing on each restart
+            Assert.Equal(typeof(SystemTraceWriter), config.TraceWriter.GetType());
```

Comment:
Before the fix, config.TraceWriter ended up being a Russian nesting doll of writers composed with writers.